vllm/v1/worker/mps_model_runner.py

```diff
@@ -0,0 +1,146 @@
+from vllm.v1.kv_cache_interface import KVCacheConfig
+from vllm.v1.worker.gpu_model_runner import GPUModelRunner
+
+logger = init_logger(__name__)
+
+
+class MPSModelRunner(GPUModelRunner):
+    """Model runner for MPS (Metal Performance Shaders) on Apple Silicon.
+
+    This runner keeps the model and input tensors on MPS for GPU acceleration,
+    but allocates the KV cache on CPU for numerical stability. The attention
+    backend handles the device movement between MPS and CPU.
+    """
+
+    def __init__(self, vllm_config: VllmConfig, device: torch.device):
+        assert device == torch.device("mps"), f"Expected MPS device, got {device}"
+
+        # Initialize with MPS device - inputs and model will be on MPS
+        with _torch_cuda_wrapper():
+            super().__init__(vllm_config, device)
+
+        assert self.speculative_config is None, "Speculative decoding is not supported on MPS."
+
+        self.use_cuda_graph = False
+        self.cascade_attn_enabled = False
+
+    def load_model(self, eep_scale_up: bool = False) -> None:
+        """Load the model onto MPS device."""
+        logger.info("Starting to load model %s...", self.model_config.model)
+
+        # Load model - it will be loaded on MPS due to device config
+        self.model = get_model(vllm_config=self.vllm_config)
+
+        # Ensure model is on MPS
+        if not all(p.device.type == "mps" for p in self.model.parameters() if p.numel() > 0):
+            logger.info("Moving model to MPS device...")
+            self.model = self.model.to("mps")
+
+        if self.lora_config:
+            self.model = self.load_lora_model(self.model, self.vllm_config, self.device)
+
+        logger.info("Model loaded successfully on MPS.")
+
+    def initialize_kv_cache_tensors(
+        self, kv_cache_config: KVCacheConfig, kernel_block_sizes: list[int]
+    ) -> dict[str, torch.Tensor]:
+        """Initialize the KV cache tensors on MPS.
+
+        Keep KV cache on MPS to avoid cross-device issues.
+        """
+        # Keep KV cache on MPS (same as model)
+        kv_caches = super().initialize_kv_cache_tensors(kv_cache_config, kernel_block_sizes)
+
+        # Log KV cache device info
+        for name, cache in kv_caches.items():
+            logger.info("KV cache %s allocated on %s", name, cache.device)
+
+        return kv_caches
+
+    def get_model(self) -> nn.Module:
+        return self.model
+
+    def warming_up_model(self) -> None:
+        """Warm up the model for compilation."""
+        logger.info("Warming up model for MPS...")
+        # Only generate graph for the generic shape
+        with _set_global_compilation_settings(self.vllm_config):
+            self._dummy_run(
+                min(
+                    max(16, self.max_num_reqs),
+                    self.scheduler_config.max_num_batched_tokens,
+                )
+            )
+        # Synchronize MPS
+        if hasattr(torch.mps, 'synchronize'):
+            torch.mps.synchronize()
```

Comment:
Is it important that this is synchronized here? We could remove this override if not.